Prolly Tree (#2030)
dolthub/docs: content/architecture/storage-engine/prolly-tree.md

A B-tree is optimized for a tradeoff between write and read performance. It's more expensive to maintain than dropping tuples into a heap without any ordering constraints, but when data is stored in a B-tree it's much quicker to seek to a given key and to do an in-order traversal of the keys with their values.

However, finding the differences between two B-trees requires scanning both trees and comparing all values, an operation that scales with the size of the tree. This operation is slow for large trees.

Also, writes to B-trees are not history independent, the order of the writes internally changes the structure of the tree. Thus, storage cannot be easily shared between two versions of the same tree.

## Building a Prolly Tree

The easiest way to understand Prolly trees is to walk through, step-by-step, how we build one. You start with a map of keys to values. Using the above B-tree example, we have keys between 1 and 21. The values don't really matter. Just imagine them dangling off the keys.

1. **Sort**: Sort the map by its key value, so that it is laid out in order. 

![Building a Prolly-Tree, Step 1](../../.gitbook/assets/tim-prolly-tree-step-1.png)

> 2. **Determine Chunk Boundaries**: Use a seed, the size of the current chunk, and the key value to calculate a rolling hash. Any time the hash value is below a target value, form a chunk boundary and start a new block. Here's the chunking step on our leaf nodes:

Reviewer: Instead of calling this a rolling hash, you might just say "Use a seed, the size of the current chunk, the key value and a strong hash function to calculate whether this current entry represents a new chunk boundary". Or something like that. It's not technically rolling at this point, just based on the key bytes.

![Building a Prolly-tree, Step 2](../../.gitbook/assets/tim-prolly-tree-step-2.png)

3. **Hash each Chunk**: You now have the leaf nodes of the Prolly Tree. Compute the content address of each block by applying a strong hash function to its contents. You store the contents in a content addressed block store. Here our blocks have been addressed and stored in the block store:

![Building a Prolly-tree, Step 3](../../.gitbook/assets/tim-prolly-tree-step-3.png)

4. **Finished?**: If the length of your list of content addresses is 1, then you are done. This is the content address of your tree.

5. **Build a New Map** Otherwise, form the next layer of your Prolly Tree by creating a map of the highest key value in the chunk and the content address of the chunk. Here's what the entries for the first internal level of the tree would look like for our example:

![Building a Prolly-tree, Step 5](../../.gitbook/assets/tim-prolly-tree-step-5.png)

> 6. **Return to Step 2**: Use this new map step 2. The algorithm terminates when step 4's condition is reached.

Reviewer: Use the sorted key value pairs for this new map as input to step 2.

Following these steps results in the following Prolly Tree. 

![Prolly Tree](../../.gitbook/assets/tim-prolly-tree-example.png)

## Modifying a Prolly Tree

The magic of Prolly Trees is not seen on construction, but on modification. As you'll see, modifying a Prolly Tree changes the minimal amount of content addresses. Minimal modification means more structural sharing of contents because fewer content addresses change. It is also the basis of fast diff which is accomplished by comparing content addresses.

### Update a Value

If we update a value, we walk the tree by key to the leaf node holding the value. We then edit the referenced chunk in place. After the edit, we recalculate the content address of the chunk. We then walk up the tree recalculating each internal content address up to the root of the tree. 

![Prolly Tree Value Update](../../.gitbook/assets/prolly-tree-update-value.png)

> 

Reviewer: Not sure if it's overly specific to mention this is only true for fixed-size value updates. If you change the length of a value, then it does have to re-chunk.

Instead of saying "in-place", it might be more clear if we said we could create the new leaf chunk by directly modifying the existing value in a copy of the existing chunk. That way it's clear that we're copy-on-write and the old chunk needs to stay around for versioning.

### Insert Keys

We can insert keys at the beginning, middle, or end of the key space. It's helpful to visualize each type of insert.

> When we insert a key, we walk the tree by key to the leaf node where the key belongs. We then edit the chunk, calculate whether to split the chunk or not, and then recalculate the content address. Note, a key insert or delete has a small probability of splitting the chunk. If the rolling hash value when incorporating the new key randomly comes up below our boundary threshold, the chunk is split at the new boundary. Finally, we walk up the tree recalculating each content address up to the root of the tree.

Reviewer: Maybe something like:

Note, a key insert or a delete has a small probability of changing an existing chunk boundary. If the computed hash values for the keys, combined with the size of chunk at those keys, happens to chose a different chunk boundary, the chunk will be split at the new boundary.

> In Dolt's Prolly Tree implementation, chunks are set to be on average 4 kilobytes. This means that you have a base probability of 1/4096 or 0.02% of triggering a chunk boundary shift with a single byte rolling hash window. A single byte window would not be enough data to provide the requisite level of randomness so the rolling hash window is set to 64 bytes, providing a base probability of a chunk split of 64/4096 or 1.6%. As we'll discuss later, the size of the chunk is considered when computing the rolling hash. So, the larger the chunk, the greater the probability the chunk will be split on the addition of a new key. Conversely, for small chunks, the probability of a chunk split is very small. This means that every edit to a table in Dolt is a minimum of 4Kb multiplied by the depth of the tree.

Reviewer: This is a little muddied because we no longer use a rolling hash.

In reality, inserting or deleting a key has some probability of changing the existing chunking because of the following factors:

* The new key might happen to have a very "good" hash, which really wants to split.
* The removed key might happen to have had a very "good" hash, which really wanted to split at it.
* Adding or removing data from the chunk will change the "target value" all other keys are looking for when they compare their hashes on the re-chunking phase.

In reality, what happens is, the existing chunk picked a key because it was both
* already big enough
* the key had a pretty "good" hash value.

In particular, the keys which came directly before the key where the existing split is – their hashes are not dramatically better than its hash and many of them are likely to be worse.

Because we chunk on a per-key bases, the probability of splitting or re-chunking is more based on the size of the data mutation you make and on the number of keys that are in the chunk in general. For example, if you tuples that are ~128 bytes then there will only be ~32 tuples per chunk, and then we do expect at least ~1/32 inserts or deletes to cause a chunk boundary change.

Author: oof

#### At the Beginning

When you insert a key at the beginning of the key space, the tree is modified along the left edge.

![Prolly Tree Beginning Insert](../../.gitbook/assets/prolly-tree-front-insert.png)

#### At the end

When you insert a key at the end of the key space, the tree is modified along the right edge.

![Prolly Tree End Insert](../../.gitbook/assets/prolly-tree-back-insert.png)

#### In the middle

When you insert a key in the middle of the key space, the tree is modified along a spline.

![Prolly Tree Middle Insert](../../.gitbook/assets/prolly-tree-middle-insert.png)

### Delete a key

When you delete a key, the tree is modified under the same rules as an insert.

![Prolly Tree Delete Key](../../.gitbook/assets/prolly-tree-delete.png)

# Properties

## History Independence

A key property of a Prolly tree that enables fast diff and structural sharing is history independence. No matter which order you insert, update, or delete values, the Prolly tree is the same. This is best seen through example.

> Consider a map with 4 integer keys, `(1, 2, 3, 4)` pointing at the same values. Before hand we know the combination of keys `(1, 2)` will trigger a chunk boundary. We can know this beforehand because chunk boundaries are based on the size of the chunk and its contents. No matter which order we insert, update, or delete, if we will end up with a tree with two leaf nodes `(1, 2)`, and `(3,4)`. This also happens to be true if the keys have different values. The tree will be the same but the the chunks will have different addresses.

Reviewer: ```suggestion
Consider a map with 4 integer keys, `(1, 2, 3, 4)` pointing at the same values. Before hand we know the combination of keys `(1, 2)` will trigger a chunk boundary. We can know this beforehand because chunk boundaries are based on the size of the chunk and its contents. No matter which order we insert, update, or delete, if we will end up with a tree with two leaf nodes `(1, 2)`, and `(3,4)`. This also happens to be true if the values at these keys are different. The tree will be the same but the the chunks will have different addresses.
```

Not sure if it's worth pointing out this is only true for same-sized values.

Let's say we insert the chunks in sequential order.

![Prolly Tree History Independence](../../.gitbook/assets/prolly-tree-history-independence-1.png)

Then, let's say we insert the keys in reverse order.

![Prolly Tree History Independence](../../.gitbook/assets/prolly-tree-history-independence-2.png)

> As you can see, we end up with the same Prolly Tree no matter which order we insert the values. It's a fun exercise to try and com e up with a sequence of inserts, updates, and deletes that result in a different tree that contains the same values. It's fun because you can't. The Prolly Tree algorithm always spits out the same tree.

Reviewer: ```suggestion
As you can see, we end up with the same Prolly Tree no matter which order we insert the values. It's a fun exercise to try and come up with a sequence of inserts, updates, and deletes that result in a different tree that contains the same values. It's fun because you can't. The Prolly Tree algorithm always spits out the same tree.
```

![Content Addressed Block Store](../../.gitbook/assets/tim-prolly-tree-step-3.png)

Thus, any blocks that share the same content address are only stored in the block store once. When they need to be retrieved they are retrieved via content address. This means that any blocks shared across versions will only be in the block store once.

As noted earlier, in Dolt's Prolly Tree implementation, the block size is set to be on average 4 kilobytes. Thus, a single change will cause a change in the block store of 4 kilobytes times the size of the tree on average.

# The Nitty Gritty Details

Now that you have the general details, let's dive into some details highlighting some of the things we learned iterating on the Noms implementation of Prolly Trees and finally landing on Dolt's stable Prolly Tree implementation.

## Controlling Chunk Size

The original Noms implementation of Prolly Trees was susceptible to a chunk size problem. You would end up with many small chunks and a few large ones. The chunk size was a geometric distribution with an average size of 4 kilobytes.

![Chunk Size Geometric](../../.gitbook/assets/chunk-distribution-static-pattern.png)

> This is the pattern one would expect from repeated rolls of a rolling hash function. Let's imagine you have a six-sided die. You walk along the road picking up pebbles. For each pebble, you put it in a bag and roll the die. If the die shows 6 you get a new bag. The average number of pebbles you have in each bag is 3.5 but you will mostly have bags with one pebble and a few bags with >10 pebbles. The pebbles in the Noms case was the key, value byte stream and the dice was the rolling hash function.

Reviewer: ```suggestion
This is the pattern one would expect from repeated rolls of a rolling hash function. Let's imagine you have a six-sided die. You walk along the road picking up pebbles. For each pebble, you put it in a bag and roll the die. If the die shows 6 you get a new bag. The average number of pebbles you have in each bag is 3.5 but you will mostly have bags with one pebble and a few bags with >10 pebbles. The pebbles in the Noms case was the key-value byte stream and the dice was the rolling hash function.
```
